common/src/mmap.rs

```diff
@@ -20,8 +21,7 @@ use rustix::shm::ShmOFlags;
 #[derive(Debug)]
 pub struct Mmap {
     fd: OwnedFd,
-    ptr: NonNull<std::ffi::c_void>,
-    len: usize,
+    buffer: NonNull<[u8]>,
```

Comment:
Doesn't this add extra indirection? We have a pointer to an array, which is a pointer and a len. There's also no point in using NonNull here, since [u8] can, by Rust's rules, never be null.

If you want to keep the indirection it's better to use a `Box<[u8]>`. Otherwise, you can keep the code as it was before, maybe just with a pointer to `u8` directly, rather than `c_void`: `NonNull<u8>`.

Reply:
Also, I believe just keeping the `NonNull<u8>` directly may be better simply because you will end up manipulating the buffer's core components anyways later in the code.

Reply:
`NonNull<T>` is defined as `*mut T` with rustc attribute for layout optimizations, so there is a reason. 

> [u8] can, by Rust's rules, never be null.

I don't understand. `[u8]`, by itself, can never exist. It must be behind a pointer.
So `&[u8]`, `*mut [u8]` and friends are basically just `struct Pointer { addr: usize, len: usize }`. I think I misunderstood.

> If you want to keep the indirection it's better to use a `Box<[u8]>`. Otherwise, you can keep the code as it was before, maybe just with a pointer to `u8` directly, rather than `c_void`: `NonNull<u8>`.

It is not. The buffer was provided by `mmap`, so on `Drop`, the `Box` would take the pointer, go to system allocator, try to deallocate and probably segfault.

I think `NonNull<[u8]>` conveys that `Mmap` is guaranteed to own slice of bytes but is unable to use `&mut [u8]` because it couldn't assign a lifetime. `'static` is incorrect, because the lifetime is however long `Mmap` will live.

Given that `Mmap`s purpose is just to be a glorified buffer, it is better to use `[u8]` as the type pointed to.

Reply:
You are right on all accounts. I am the one who got confused, my apologies.

I also agree with your arguments on why to use `[u8]` rather than `u8`.